### DISCUSSION TITLE
language configuration objects were removed in v36

### DISCUSSION THREAD

**Discussion by yz6028693** [renovatebot/renovate#24210] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

version 36.25.0

### If you're self-hosting Renovate, select which platform you are using.

Azure DevOps (dev.azure.com)

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Wanted end result.

We have set the private registry in hostRules, and set rule like:

"python": {
                "registryUrls": [
                                        "our_company_registry_url/{feed_name}/pypi/simple/"
                                       ]
                 },

This setting works fine when we use renovate version 35, but renovate started to hit public pypi registry instead of our private registry after we migrate to version 36.25.0.

### What you tried so far.

There is no index-url in our python requirement.txt file

Adding packageRules below seems like a walk around and can let renovate hit private registry. But may I know why the previous setting won't work again?

{
              "matchDatasources": ["pypi"],
              "registryUrls": [
                "our_company_registry_url/{feed_name}/pypi/simple/"
              ]
},

### Relevant debug logs

<details><summary>Logs</summary>

```
                                                     "hostRules": [
2023-08-22T20:49:08.1387823Z            {"enabled": false},
2023-08-22T20:49:08.1391764Z            {
2023-08-22T20:49:08.1391864Z              "matchHost": "our_company_registry_url",
2023-08-22T20:49:08.1391973Z              "username": "our_user_name",
2023-08-22T20:49:08.1392071Z              "password": "our_password",
2023-08-22T20:49:08.1392169Z              "enabled": true
2023-08-22T20:49:08.1392246Z            },
2023-08-22T20:49:08.1392885Z          ],

2023-08-22T20:49:08.1463411Z          "python": {
2023-08-22T20:49:08.1463502Z            "registryUrls": [
2023-08-22T20:49:08.1463717Z              "our_company_registry_url/{feed_name}/pypi/simple/"
2023-08-22T20:49:08.1463848Z            ]
2023-08-22T20:49:08.1463924Z          },


2023-08-22T20:49:13.8589923Z DEBUG: Host is disabled - rejecting request. HostUrl: https://pypi.org/pypi/apscheduler/json
2023-08-22T20:49:13.8595239Z DEBUG: Failed to look up pypi package apscheduler


2023-08-22T20:49:18.4375870Z              "deps": [
2023-08-22T20:49:18.4375950Z                {
2023-08-22T20:49:18.4376032Z                  "depName": "apscheduler",
2023-08-22T20:49:18.4376134Z                  "currentValue": "==3.10.3",
2023-08-22T20:49:18.4376233Z                  "datasource": "pypi",
2023-08-22T20:49:18.4376326Z                  "currentVersion": "3.10.3",
2023-08-22T20:49:18.4376422Z                  "updates": [],
2023-08-22T20:49:18.4376516Z                  "packageName": "apscheduler",
2023-08-22T20:49:18.4376653Z                  "versioning": "pep440",
2023-08-22T20:49:18.4376747Z                  "warnings": [
2023-08-22T20:49:18.4376824Z                    {
2023-08-22T20:49:18.4376911Z                      "topic": "apscheduler",
2023-08-22T20:49:18.4377026Z                      "message": "Failed to look up pypi package apscheduler"
2023-08-22T20:49:18.4377157Z                    }
2023-08-22T20:49:18.4377235Z                  ]
2023-08-22T20:49:18.4377307Z                },


2023-08-22T20:49:08.1674140Z DEBUG: Setting global hostRules
2023-08-22T20:49:08.1695139Z DEBUG: Adding password authentication for our_company_registry_url to hostRules
2023-08-22T20:49:08.1698490Z DEBUG: Adding token authentication for dev.azure.com to hostRules
2023-08-22T20:49:08.1707650Z DEBUG: validatePresets()
2023-08-22T20:49:08.5808936Z DEBUG: Reinitializing hostRules for repo
2023-08-22T20:49:08.5809235Z DEBUG: Clearing hostRules
2023-08-22T20:49:08.5811831Z DEBUG: Adding password authentication for our_company_registry_url to host
```

</details>

**Comment by yz6028693** [comment]
Wondering if it's because the team removed the "python" object in version 36 (e.g.: in lib/config/options/index.ts)

  {
    name: 'python',
    description: 'Configuration object for Python.',
    stage: 'package',
    type: 'object',
    default: {},
    mergeable: true,
    cli: false,
  },
  
  I found using pip_requirements object will work, but what would be the best option here (to set private registry for all python pkgs)
  
"pip_requirements": {
"registryUrls": [
"our_company_registry_url/{feed_name}/pypi/simple/"
]
},

**Reply by rarkins** [reply]
Try "pypi" instead. And also please update the title of this discussion as it's very misleading for other users

**Reply by yz6028693** [reply]
Thanks Rarkins! Just updated the title, will test pypi next week!

**Reply by yz6028693** [reply]
Hi @rarkins, I just tried to use "pypi", but seems like it won't work and renovate will point to public pypi registry with using it. Any ideas?

Version of the renovate: 36.0.0; The setting I have:

"pypi": {
    "registryUrls": [
        "our_company_registry_url/{feed_name}/pypi/simple/"
    ]
},

On the other hand setting below will work:

"pip_requirements": {
    "registryUrls": [
        "our_company_registry_url/{feed_name}/pypi/simple/"
    ]
},

Our python pkgs are listed in a python requirements.txt file.

Please kindly let me know if any logs need to be provided

**Reply by rarkins** [reply]
Can you try this instead?

```
"packageRules": [{
  "matchDatasources": ["pypi"],
  "registryUrls": [
    "our_company_registry_url/{feed_name}/pypi/simple/"
  ]
}]
```

The `pip_requirements` approach is also ok, but the idea behind a `pypi` rule is that it would apply to all managers at once (pipenv, poetry, pip_requirements, etc)

**Reply by acortelyou** [reply]
> Try "pypi" instead.

The [documentation](https://docs.renovatebot.com/python/#specify-url-in-configuration) still suggests to use `python` config object in order to configure the default `registryUrls` for python packages. Does this need to be updated?

> Can you try this instead? `packageRules.[{matchDatasources:pypi}]`

My understanding is that a packageRule would **override** the effect of any `index-url` found in a `requirements.txt`.

The behavior which is desired, and was seen 35, is to set the **default** registryUrl for python-category managers but still allow `index-url` in a `requirements.txt` to work as intended.

This nuance is described in the same doc as above:
> If a requirements.txt file has an index-url then Renovate follows that link, instead of following any link set in the [python.]registryUrls array. To **override** the URL found in requirements.txt, you must create a custom packageRules setting. **This is because packageRules are applied after package file extraction.**


### To summarize my understanding:

1) [language config objects were removed causing breaking behavior changes](https://github.com/renovatebot/renovate/commit/19259a03834c38ee81eda915b20cfea21b4bc1e5#diff-0053f214ec552bf58e21e52b0af7fcfe1699c7be2989cee50c52d44c7046ed2fL2125-L2182)
2) the python docs (and possibly more) need to be updated to reflect those changes
3) ~~packageRules for matchCategories/matchDatasource are not enough to emulate previous behavior~~ (see below)
4) ~~previous behavior should be achievable by configuring the `pypi` datasource, **but this doesn't work for some reason**~~ (datasources are not configurable directly)
5) previous behavior is achievable by configuring the `pip_requirements` manager, **but there isn't a way to configure all python-category managers together**
6) the categories feature changes included packageRules migrations for `matchLanguage`=>`matchCategories`, **but does not appear to include config objects migrations for `language`=>`category`**

### Open questions:

1. **Why doesn't `config.pypi.registryUrls` have the expected effect** (answer: datasources are not configurable like managers)
2. Which of the following, if any, would be useful to consider:
  a. `category` config objects which apply to all the matching managers at once (answer: managers are best configured explicitly)
  b. migration from `language` config object to the corresponding `datasource` config objects (answer: a more clear explanation in the major release changlog would have been nice, or a helpful `WARN`)


nice work @yz6028693 for tracking down this regression, thanks!

**Reply by rarkins** [reply, marked as answer]
Relevant: https://docs.renovatebot.com/configuration-options/#defaultregistryurls

**Reply by acortelyou** [reply]
**fantastic!**

**thanks @rarkins for pointing that out.**

i think the confusion stemmed from:
* outdated python language support docs (`config.python` block no longer does anything, needs to be updated to use `packageRules`)
* datasources aren't configurable in the same way managers are (`config.pypi` block doesn't do anything)
* `nuget`, `maven` and `npm` are all both datasources and managers
* conflating the now obsolete `language` config with `datasources` instead of `managers`
* completely overlooking the existence of `defaultRegistryUrls`

before and after config snippets in case someone needs this in the future, please forgive my yaml-ese

#### **`renovate-config-only-seemed-to-work-in-35.yaml`**
```yaml
hostRules:
- enabled: false
- matchHost: https://pkgs.dev.azure.com
  enabled: true

python:
  registryUrls:
  - https://${{feed}}/pypi/simple/
maven:
  registryUrls:
  - https://${{feed}}/maven/v1
npm:
  registryUrls:
  - https://${{feed}}/npm/registry/
nuget:
  registryUrls:
  - https://${{feed}}/nuget/v3/index.json
```
#### **`renovate-config-actually-works-in-35-and-36.yaml`**
```yaml
hostRules:
- enabled: false
- matchHost: https://pkgs.dev.azure.com
  enabled: true

packageRules:
- matchDatasources: ["pypi"]
  defaultRegistryUrls:
  - https://${{feed}}/pypi/simple/
- matchDatasources: ["maven"]
  defaultRegistryUrls:
  - https://${{feed}}/maven/v1
- matchDatasources: ["npm"]
  defaultRegistryUrls:
  - https://${{feed}}/npm/registry/
- matchDatasources: ["nuget"]
  defaultRegistryUrls:
  - https://${{feed}}/nuget/v3/index.json
```

**Reply by rarkins** [reply]
Thanks for sharing!
